Support `USE catalog.schema` statements

mdesmet opened the pull request:
Fixes #153

ebyhr commented on tests/integration/test_dbapi_integration.py:
```diff
@@ -859,3 +859,13 @@ def test_info_uri(trino_connection):
+
+def test_use_schema(trino_connection):
+    cur = trino_connection.cursor()
+    cur.execute('use tpch.tiny')
```
nit: Uppercase `use`

ebyhr commented on tests/integration/test_dbapi_integration.py:
```diff
@@ -859,3 +859,13 @@ def test_info_uri(trino_connection):
+
+
+def test_use_schema(trino_connection):
+    cur = trino_connection.cursor()
```
Please add assertions to verify the original catalog and schema before executing `USE` query.

hashhar replied:
Should we also assert that new cursors opened after a `USE` still use the default catalog and schema specified on the connection?

mdesmet replied:
Currently it seems indeed that new cursors don't take over the session properties.

However I don't think this will match the expectation of the users:

> JDBC defines a system session and transaction model for database access. A session is the duration of one connection to the database and is handled by a JDBC Connection object.

See [oracle docs](https://docs.oracle.com/javadb/10.8.3.0/devguide/cdevconcepts38375.html#:~:text=JDBC%20defines%20a%20system%20session,by%20a%20JDBC%20Connection%20object.&text=A%20Connection%20object%20represents%20a%20connection%20with%20a%20database.)

WDYT?

hashhar replied:
DBAPI != JDBC. We should do what DB-API says and if it doesn't say explicitly then follow what some other drivers do (psycopg for example doesn't maintain search path changes across cursors last time I checked).

mdesmet replied:
I checked the DBAPI specs, nothing mentioned about this. IMHO it's more of a general expectation. See e.g. https://stackoverflow.com/questions/8797724/mysql-concepts-session-vs-connection

Snowflake for example does apply it to the connection:

https://github.com/snowflakedb/snowflake-connector-python/blob/107c3033cb2c7c6961d06aa613e7c3a8f20ae435/src/snowflake/connector/cursor.py#L753-L759

In a lot of cases this is also not really a client thing, but something stored within the database side connection/session.

mdesmet replied:
@ebyhr , @hashhar : WDYT?

mdesmet replied:
@hovaesco : What's your view on this? I think in the PR's code we currently do some locking in `ClientSession`. IMHO this should be per connection instead of per cursor.

hovaesco replied:
It would be more useful when doing locking on connection/sesion level not per cursor which is limited and not user-friendly.

mdesmet commented:
Not sure why the build failed in 3.10 on the following error:

```
trino.exceptions.TrinoUserError: TrinoUserError(type=USER_ERROR, name=COLUMN_NOT_FOUND, message="line 1:8: Column 'current_catalog' cannot be resolved", query_id=20220424_155121_00213_82y8r)

```

ebyhr commented:
`current_catalog` was added in version 357, so it fails on 351.
https://trino.io/docs/current/release/release-357.html

hovaesco reviewed (approved):
LGTM

ebyhr commented on tests/integration/test_dbapi_integration.py:
```diff
@@ -859,3 +859,25 @@ def test_info_uri(trino_connection):
+    assert result[0][0] is None
+    assert result[0][1] is None
+
+    cur.execute('USE tpch.tiny')
```
Let's add another test case to specify only schema name. e.g. `USE sf1`.

ebyhr reviewed (approved):
Could you please rebase on upstream to resolve conflicts?

mdesmet commented:
I have rebased the code and made following changes:

- `ClientSession` is now thread-safe. it is initiated by the `dbapi.Connection` and some of its properties can be modified based on the headers coming back from the Trino server. It can later on be extended with more modifiable properties. 
- Added tests for both `USE catalog.schema` and `USE schema`
- Improved mypy typing for `ClientSession` class

hovaesco reviewed (commented):
Could you please (if possible) split the changes to two commits: refactor of `ClientSession` and add `USE` functionality?

ebyhr commented:
Merged, thanks!
